### DISCUSSION TITLE
using @Brick "StandardElasticity"

### DISCUSSION THREAD

**Discussion by GhiathMonnet** [thelfer/tfel#381] [General]
Dear colleagues,
I would like to use @Brick "StandardElasticity". I tried to introduce the values of Young Modulus and Poisson's ratio in mtest. No matter how I call these modulii : YoungModulus, youngmodulus, Young, etc.. I always get an error message teeling me that the law does not declare such variable.
I am using the "generic" interface.
All examples I have seen are introducing the modulii inside the mfront file itself and not in mtest.
Many thanks for your help.
Cheers

**Comment by thelfer** [comment]
You may have a look at the documentation of the `StandardElasticity` brick for details about how elastic material properties are handled: https://thelfer.github.io/tfel/web/HookeStressPotential.html

Assuming that your behaviour is isotropic, the brick automatically declares the `YoungModulus` and `PoissonRatio` as material properties if none of the following conditions is met (those conditions are exclusive):

- The elastic properties are defined in the options to the brick, as follows:
  ```cxx
  @Brick StandardElasticity{
      young_modulus: 150e9,
      poisson_ratio: 0.3
  };
  ```
- The elastic properties are defined by the keyword `@ElasticMaterialProperties`
- The stiffness tensor is defined by one of the keyword `@RequireStiffnessTensor` or `@ComputeStiffnessTensor`
-  Two material properties or parameters are defined with the external name `YoungModulus` or `PoissonRatio`, as follows:
    ```cxx
    @MaterialProperty stress yg;
    yg.setGlossaryName("YoungModulus");
    @MaterialProperty stress nu;
    nu.setGlossaryName("PoissonRatio");
     ```
     
  See the attached example and the following code:
  
  ```
  $ mfront-query --material-properties StandardElasticityTest.mfront 
- YoungModulus (young): the Young's modulus of an isotropic material
- PoissonRatio (nu): the Poisson ratio of an isotropic material
  ```

[StandardElasticityTest.zip](https://github.com/thelfer/tfel/files/11110953/StandardElasticityTest.zip)

**Comment by GhiathMonnet** [comment]
Thomas : Many thanks for this detailed information. OK, then  I have to inset in the mtest file the following lines:
@MaterialProperty 'YoungModulus' 200e9  ;
@MaterialProperty 'PoissonRatio' 0.3 ;
 
Cheers.
Ghiath

**Reply by thelfer** [reply]
In this case that I sent you, yes.

**Comment by GhiathMonnet** [comment]
Dear Thomas. I tried this but it did not work.

As you suggest, I declared in mtest :

```cxx
@MaterialProperty<function> 'YoungModulus' "2.6e6 * (-23.721 * Temperature + 78821)" ;
@MaterialProperty<constant> 'PoissonRatio' 0.3 ;
```

In the mfront file, I put  : 

```cxx
@DSL Implicit;
@Behaviour Monnet_RPV_EVP;
@Algorithm NewtonRaphson_NumericalJacobian ;
@Description{
}
@Epsilon 1.e-14;
@Theta 1;
@Brick "StandardElasticity"; 
// when i want to compute the shear modulus as follows
@InitLocalVariables {
    mu = YoungModulus/2./(1.+ PoissonRatio);              // module de cisaillement
```

I have the error message :  `error: ‘YoungModulus’ was not declared in this scope`

I tried `mu = computeMu(YoungModulus,PoissonRatio)`. I have the same error.

Can you explain to me how can I use :  `@Brick "StandardElasticity`, then provide the value of `YoungModulus` and `PoissonRatio` in `mtest`, and then compute the shear modulus in the `mfront` file ? 

Thanks in advance.
Ghiath

**Reply by thelfer** [reply]
Hi Ghiath,

You seem to be confused by the concept of external names in `MFront`. In `MFront`, a variable `x` is associated with an external name by the `setGlossaryName` and `setEntryName` methods.. `x` is the variable that you manipulate in `MFront` and its external name is what you see from the solver (`MTest` in you case).

As mentionned in the documentation of the brick: (https://thelfer.github.io/tfel/web/HookeStressPotential.html), by **default**, the brick modulus will automatically define the Young' modulus  and the Poisson's ratio as material properties named `young` and `nu` associated respectively with the external names `YoungModulus` and `PoissonRatio`

So you can compute the shear modulus in `MFront` in your example as:

```cxx
mu = computeMu(young, nu);
```

but you'll have to define the material properties `YoungModulus` and `PoissonRatio` in your `MTest` file.

**Comment by GhiathMonnet** [comment]
Many Thanks.
The explanation in  (https://thelfer.github.io/tfel/web/HookeStressPotential.html) is not clear for me. 
What that means "if the stiffness tensor is available" ? If the solver does not provide the elastic modulii (through mtest for example), how the stiffness tensor can be available ? Who is providing the stiffness tensor when it is "available" ?

**Reply by thelfer** [reply]
The previous paragraph states:

> If the elastic behaviour is orthotropic, the stiffness tensor must be available (using the keyword
> @RequireStiffnessTensor) or computed by the behaviour (using the keyword @ComputeStiffnessTensor). 

See the documentation of those keywords

The effect of the  @RequireStiffnessTensor keywords depends on the interface. It generally works by adding additional material properties. The material properties are not accessible from the `MFront` file, you only have access to the  stiffness tensor D.

**Comment by GhiathMonnet** [comment]
many thanks. Thomas.
